utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,414 @@
+
+data EntryPoint = EntryPoint
+  { name :: Text
+  , mode :: ShaderStage
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data Input = Input
+  { type' :: Text
+  , name :: Text
+  , location :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic)
+
+instance FromJSON Input where
+  parseJSON = withObject "inputs" $ \v -> Input
+    <$> v .: "type"
+    <*> v .: "name"
+    <*> v .: "location"
+    <*> v .:? "array"
+
+instance ToJSON Input where
+  toJSON (Input type' name location array) = object
+    [ "type" .= type'
+    , "name" .= name
+    , "location" .= location
+    , "array" .= array
+    ]
+  toEncoding (Input type' name location array) = pairs
+    (  "type" .= type'
+    <> "name" .= name
+    <> "location" .= location
+    <> "array" .= array
+    )
+
+data Ubo = Ubo
+  { name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data Texture = Texture
+  { type' :: Text
+  , name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic)
+
+instance FromJSON Texture where
+  parseJSON = withObject "textures" $ \v -> Texture
+    <$> v .: "type"
+    <*> v .: "name"
+    <*> v .: "set"
+    <*> v .: "binding"
+    <*> v .:? "array"
+
+instance ToJSON Texture where
+  toJSON (Texture type' name set binding array) = object
+    [ "type" .= type'
+    , "name" .= name
+    , "set" .= set
+    , "binding" .= binding
+    , "array" .= array
+    ]
+  toEncoding (Texture type' name set binding array) = pairs
+    (  "type" .= type'
+    <> "name" .= name
+    <> "set" .= set
+    <> "binding" .= binding
+    <> "array" .= array
+    )
+
+data Reflection = Reflection
+  { entryPoints :: Vector EntryPoint
+  , inputs :: Maybe (Vector Input)
+  , textures :: Maybe (Vector Texture)
+  , ubos :: Maybe (Vector Ubo)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data ShaderStage
+  = Vert
+  | Frag
+  | Comp
+  | Tesc
+  | Tese
+  | Geom
+  | Rgen
+  | Rint
+  | Rahit
+  | Rchit
+  | Rmiss
+  | Rcall
+  | Task
+  | Mesh
+  deriving (Eq)
+
+instance IsString ShaderStage where
+  fromString = \case
+    "vert" -> Vert
+    "frag" -> Frag
+    "comp" -> Comp
+    "tesc" -> Tesc
+    "tese" -> Tese
+    "geom" -> Geom
+    "rgen" -> Rgen
+    "rint" -> Rint
+    "rahit" -> Rahit
+    "rchit" -> Rchit
+    "rmiss" -> Rmiss
+    "rcall" -> Rcall
+    "task" -> Task
+    "mesh" -> Mesh
+    unsupport -> error $ "ShaderStage not support '" <> unsupport <> "'"
+
+instance Show ShaderStage where
+  show = \case
+    Vert -> "vert"
+    Frag -> "frag"
+    Comp -> "comp"
+    Tesc -> "tesc"
+    Tese -> "tese"
+    Geom -> "geom"
+    Rgen -> "rgen"
+    Rint -> "rint"
+    Rahit -> "rahit"
+    Rchit -> "rchit"
+    Rmiss -> "rmiss"
+    Rcall -> "rcall"
+    Task -> "task"
+    Mesh -> "mesh"
+
+instance FromJSON ShaderStage where
+  parseJSON = withText "mode" (pure . fromString . T.unpack)
```

Comment:
This should use custom conversion function instead of relying on non-total `fromString`.

Basically, don't use Show/Read/IsString for parsing. Derive show, and leave all that stuff to debug dumps.

The code will be safer and faster when using pure functions and effect-appropriate procedures. The `FromJSON` class can handle failures by itself, so you can report unsupported stages there instead of crashing user's program.

Same goes for `ToJSON`. Instead of breaking all the haskell value prettyprinters with custom Show instance, use case in `toJSON` to output the needed text values.

As a bonus, doing this will liberate you from tying constructor names to stage labels. This way you can write a proper `RayMiss` instead of `rmiss`.

Reply:
thanks for the suggestion.

- [x] custom conversion
- [x] not use Show/Read/IsString for parsing. Drive show
- [x] handle failures in `FromJSON`
- [x] use custom conversion function instead of `Show` when implement `ToJSON`
- [ ] bonus, write `RayMiss` instead of `rmiss`

what is the effect-appropriate procedures? and how to use it to make code safer and faster?

do you mean use the `RayMiss` constructor instead of `Rmiss` constructor? or use the `RayMiss` as the result of `show Rmiss`

Reply:
> what is the effect-appropriate procedures?

For example, `parseFail` from aeson when parsing JSON.

> do you mean use the `RayMiss` constructor instead of `Rmiss` constructor?

Yeah, this.

Reply:
which is better for naming shader stage constructors?

1. [refer to shaderc](https://github.com/google/shaderc/blob/011139094ec790ff7f32ea2d80286255fc9ed18b/glslc/src/shader_stage.cc#L33) 
``` haskell
data ShaderStage
  = VertexShader
  | FragmentShader
  | ComputeShader
  | TessControlShader
  | TessEvaluationShader
  | GeometryShader
  | RaygenShader  -- maybe RayGenerationShader is better?
  | RayIntersectionShader
  | RayAnyhitShader
  | RayClosesthitShader
  | RayMissShader
  | RayCallableShader
  | TaskShader
  | MeshShader
```

2. [refer to shaderc](https://github.com/google/shaderc/blob/011139094ec790ff7f32ea2d80286255fc9ed18b/glslc/src/shader_stage.cc#L33), remove `Shader` suffix
``` haskell
data ShaderStage
  = Vertex
  | Fragment
  | Compute
  | TessControl
  | TessEvaluation
  | Geometry
  | Raygen  -- maybe RayGeneration is better?
  | RayIntersection
  | RayAnyhit
  | RayClosesthit
  | RayMiss
  | RayCallable
  | Task
  | Mesh
```

3. only change ray tracing stages
``` haskell
data ShaderStage
  = Vert
  | Frag
  | Comp
  | Tesc
  | Tese
  | Geom
  | RayGen
  | RayInt
  | RayAhit
  | RayChit
  | RayMiss
  | RayCall
  | Task
  | Mesh
```

4. other

[for futher reference](https://github.com/expipiplus1/vulkan/pull/263#issue-574789938)

Reply:
2 looks good.

Expanding Raygen offers not big enough improvement to make an exception from `shaderc` reference.

Reply:
actually `RayIntersectionShader, RayAnyhitShader, RayClosesthitShader, RayMissShader, RayCallableShader` are all exceptions from `shaderc` due to their `Ray` prefix

Reply:
should we keep the `Ray` prefix?

Reply:
I checked out shaderc source and vulkan docs and... eww :confused: 

I think we should stick with [vulkan spec](https://hackage.haskell.org/package/vulkan-3.10.1/docs/Vulkan-Core10-Pipeline.html#t:ShaderStageFlagBits) with minimal deviation from stage flag naming. That is, dropping `…_STAGE` prefix and `_BIT…` suffix, and going CamelCase for multiple words that remain:

`VK_SHADER_STAGE_TESSELLATION_CONTROL_BIT` -> `TesellationControl`

That gives us `Raygen` (one word in spec), `Task` (without Ray), and `ClosestHit` (two words without ray).

Reply:
renamed.